src/pools.jl

```diff
@@ -0,0 +1,152 @@
+`release(pool)` will return the "permit" to the pool while not returning
+any object for reuse.
+
+`drain!` can be used to remove any cached objects for reuse, but it does *not* release
+any active acquires.
+"""
+mutable struct Pool{K, T}
+    lock::Threads.Condition
+    max::Int
+    cur::Int
+    keyedvalues::Dict{K, Vector{T}}
+    values::Vector{T}
+
+    function Pool{K, T}(max::Int=4096) where {K, T}
+        T === Nothing && throw(ArgumentError("Pool type can not be `Nothing`"))
+        x = new(Threads.Condition(), max, 0)
+        if K === Nothing
+            x.values = T[]
+            safesizehint!(x.values, max)
+        else
+            x.keyedvalues = Dict{K, Vector{T}}()
+        end
+        return x
+    end
+end
+
+Pool{T}(max::Int=4096) where {T} = Pool{Nothing, T}(max)
+
+safesizehint!(x, n) = sizehint!(x, min(4096, n))
+
+# determines whether we'll look up object caches in .keyedvalues or .values
+iskeyed(::Pool{K}) where {K} = K !== Nothing
+
+"""
+    drain!(pool)
+
+Remove all objects from the pool for reuse, but do not release any active acquires.
+"""
+function drain!(pool::Pool{K}) where {K}
+    Base.@lock pool.lock begin
+        if iskeyed(pool)
+            for objs in values(pool.keyedvalues)
+                empty!(objs)
+            end
+        else
+            empty!(pool.values)
+        end
+    end
+end
+
+# in VERSION >= v"1.7", we can replace `TRUE` with `Returns(true)`
+TRUE(x) = true
+
+@noinline keyerror(key, K) = throw(ArgumentError("invalid key `$key` provided for pool key type $K"))
+@noinline releaseerror() = throw(ArgumentError("cannot release permit when pool is empty"))
+
+# NOTE: assumes you have the lock!
+function releasepermit(pool::Pool)
+    pool.cur > 0 || releaseerror()
+    pool.cur -= 1
+    notify(pool.lock; all=false)
+    return
+end
+
+"""
+    acquire(f, pool::Pool{K, T}, [key::K]; forcenew::Bool=false, isvalid::Function) -> T
+
+Get an object from a `pool`, optionally keyed by the provided `key`.
+The provided function `f` must create a new object instance of type `T`.
+Each `acquire` call MUST be matched by exactly one `release` call.
+The `forcenew` keyword argument can be used to force the creation of a new object, ignoring any existing objects in the pool.
+The `isvalid` keyword argument can be used to specify a function that will be called to determine if an object is still valid
+for reuse. By default, all objects are considered valid.
+If there are no objects available for reuse, `f` will be called to create a new object.
+If the pool is already at its maximum capacity, `acquire` will block until an object is returned to the pool via `release`.
+"""
+function Base.acquire(f, pool::Pool{K, T}, key=nothing; forcenew::Bool=false, isvalid::Function=TRUE) where {K, T}
+    key isa K || keyerror(key, K)
+    Base.@lock pool.lock begin
+        # first get a permit
+        while pool.cur >= pool.max
```

Comment:
Let's assume noone holds the lock, but task (1) is waiting on the `pool.lock` condition. Is it racy when task (2) releases from the pool, just as task (3) also tries to acquire? Task 1 would wake up, task 3 would hold the lock and both of them might now pass the  `pool.cur >= pool.max` check?

Reply:
> Task 1 would wake up, task 3 would hold the lock

This isn't possible. The design of `Threads.Condition` means that when you "wake up", you are guaranteed exclusive access to the lock.